### PR TITLE
vmspawn: implement machinectl shell for vm class machines and send termination signal via D-Bus-over-SSH

### PR DESCRIPTION
**Pull request by bluca**
Cannot push to org, so continuing here from https://github.com/systemd/systemd/pull/32208

### CODE REVIEW

**Comment by evverx** [comment]
Can I ask why instead of stabilizing the release and addressing known regressions like https://github.com/systemd/systemd/issues/30421 introduced by sloppy merges new features with no tests are being pushed in a hurry?

**Comment by YHNdnzj** [comment]
Yeah, since we've already taken over the PR, there's genuinely no need to rush it in. I'll move the milestone to v257.

**Comment by bluca** [comment]
> Yeah, since we've already taken over the PR, there's genuinely no need to rush it in. I'll move the milestone to v257.

See yesterday's chat

**Comment by YHNdnzj** [comment]
> > Yeah, since we've already taken over the PR, there's genuinely no need to rush it in. I'll move the milestone to v257.
> 
> See yesterday's chat

I've read that. But there's no clear indication that we need to get this in for v256, and when you asked "anyone disagrees" I wasn't there...

**Comment by DaanDeMeyer** [comment]
This breaks registering vmspawn vms when using newer vmspawn with older machined, which is pretty much every vmspawn invocation right now. The varlink method will not be available and registering will fail.

File: src/shared/varlink-io.systemd.Machine.c [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "varlink-io.systemd.Machine.h"
+#include "varlink-idl.h"
```
Comment (YHNdnzj) [inline src/shared/varlink-io.systemd.Machine.c]: Not sorted.

File: src/vmspawn/vmspawn.c [hunk 2]
```diff
@@ -1867,6 +1861,17 @@ static int run_virtual_machine(int kvm_device_fd, int vhost_device_fd) {
                 r = unit_name_to_prefix(trans_scope, &scope_prefix);
                 if (r < 0)
                         return log_error_errno(r, "Failed to strip .scope suffix from scope: %m");
+
+                /* on distros that provide their own sshd@.service file we need to provide a dropin which
+                 * picks up our public key credential */
+                r = machine_credential_set(&arg_credentials,
```
Comment (YHNdnzj) [inline src/vmspawn/vmspawn.c]: break line before `&arg_credentials`?

File: src/vmspawn/vmspawn.c [hunk 3]
```diff
@@ -1867,6 +1861,17 @@ static int run_virtual_machine(int kvm_device_fd, int vhost_device_fd) {
...
+                                "systemd.unit-dropin.sshd-vsock@.service:"
+                                "[Service]\n"
+                                "ExecStart=\n"
+                                "ExecStart=/usr/sbin/sshd -i -o 'AuthorizedKeysFile=${CREDENTIALS_DIRECTORY}/ssh.ephemeral-authorized_keys-all .ssh/authorized_keys'\n"
```
Comment (YHNdnzj) [inline src/vmspawn/vmspawn.c]:
```suggestion
                                "ExecStart=/usr/sbin/sshd -i -o 'AuthorizedKeysFile=%d/ssh.ephemeral-authorized_keys-all .ssh/authorized_keys'\n"
```

File: src/shared/json.c [hunk 4]
```diff
@@ -5006,6 +5007,28 @@ int json_dispatch_user_group_name(const char *name, JsonVariant *variant, JsonDi
...
+
+        assert(variant);
+
+        if (!json_variant_is_string(variant))
+                return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' is not a string.", strna(name));
+
+        path = json_variant_string(variant);
+        if (!path_is_valid(path))
+                return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' is not a valid path.", strna(name));
+
```
Comment (YHNdnzj) [inline src/shared/json.c]: please drop the empty line

**Comment by bluca** [comment]
Rebased on top of https://github.com/systemd/systemd/pull/32709

**Comment by evverx** [comment]
> See yesterday's chat

Can anyone copy-paste that here for mere mortals to see?

**Comment by evverx** [comment]
Either way I guess the question isn't going to be answered. It answers my question too in a different way.

**Comment by YHNdnzj** [comment]
> > See yesterday's chat
> 
> Can anyone copy-paste that here for mere mortals to see?

Basically: this is a contract work from STF, and payment requires all the PRs to be merged.

I think there should have been a better way to communicate that these are optional and the payment should be done nonetheless, but realistically I'm never involved in all these contract works so I can't tell.

**Comment by evverx** [comment]
@bluca can you please stop deleting comments? That's not how open source projects are supposed to operate.

I've found the feedback form on the STF site and I'll contact them. I was going to do that anyway in the context of their bug resilience program. Thanks to the recent improvements (which I have no problem with) and miscommunications (which are problematic) I pulled out my downstream fuzzing infrastructure (even though it isn't broken yet) and they'd probably be interested to see the spreadsheets showing how much they should invest in the infrastructure alone to even cover the basics. Just to be clear I don't think it's STF fault. They were probably be misled into thinking that there is some actual upstream infrastructure where use-after-frees appear out of thin air.

**Comment by bluca** [comment]
Rebased and picked changes from https://github.com/systemd/systemd/pull/32482

File: src/vmspawn/vmspawn.c [hunk 5]
```diff
@@ -1919,6 +1913,17 @@ static int run_virtual_machine(int kvm_device_fd, int vhost_device_fd) {
                 log_debug("Executing: %s", joined);
         }
 
+        if (arg_register) {
+                _cleanup_free_ char *vm_address = NULL;
+
+                if (child_cid != VMADDR_CID_ANY && asprintf(&vm_address, "vsock/%u", child_cid) < 0)
+                        return log_oom();
```
Comment (keszybz) [inline src/vmspawn/vmspawn.c]:
```C
char vm_address[STRLEN("vsock/") + DECIMAL_STR_MAX(unsigned)];
xsprintf(vm_address, "vsock/%u", child_cid);
```

File: src/vmspawn/vmspawn-register.c [hunk 6]
```diff
@@ -5,36 +5,45 @@
...
-                        &error,
-                        NULL,
-                        "sayssus",
-                        machine_name,
-                        SD_BUS_MESSAGE_APPEND_ID128(uuid),
-                        service,
-                        "vm",
-                        (uint32_t) getpid_cached(),
-                        strempty(directory));
+        r = varlink_connect_address(&vl, "/run/systemd/machine/io.systemd.Machine");
```
Comment (keszybz) [inline src/vmspawn/vmspawn-register.c]: Please explain in the commit message why a D-bus call is replaced by a varlink call.

File: src/vmspawn/vmspawn.c [hunk 7]
```diff
@@ -1861,6 +1861,18 @@ static int run_virtual_machine(int kvm_device_fd, int vhost_device_fd) {
...
                         return log_error_errno(r, "Failed to strip .scope suffix from scope: %m");
+
+                /* on distros that provide their own sshd@.service file we need to provide a dropin which
+                 * picks up our public key credential */
+                r = machine_credential_set(
+                                &arg_credentials,
+                                "systemd.unit-dropin.sshd-vsock@.service:"
+                                "[Service]\n"
+                                "ExecStart=\n"
+                                "ExecStart=/usr/sbin/sshd -i -o 'AuthorizedKeysFile=%d/ssh.ephemeral-authorized_keys-all .ssh/authorized_keys'\n"
```
Comment (keszybz) [inline src/vmspawn/vmspawn.c]: Maybe drop `/usr/sbin/`? Split-bin is deprecated.

File: src/vmspawn/vmspawn.c [hunk 8]
```diff
@@ -697,6 +700,53 @@ static int setup_notify_parent(sd_event *event, int fd, int *exit_status, sd_eve
         return 0;
 }
 
+static int bus_open_in_machine(sd_bus **ret, unsigned cid, unsigned port, const char *private_key_path) {
```
Comment (keszybz) [inline src/vmspawn/vmspawn.c:710]: This is `static` and unused, so the compilation will at least cause a warning. Please merge this the commit that adds the use.

File: src/vmspawn/vmspawn.c [hunk 9]
```diff
@@ -751,7 +758,7 @@ static int on_orderly_shutdown(sd_event_source *s, const struct signalfd_siginfo
         PidRef *pidref = userdata;
         int r;
 
-        /* TODO: actually talk to qemu and ask the guest to shutdown here */
+        /* Backup method to shutdown the VM when D-BUS access over SSH is not available */
```
Comment (keszybz) [inline src/vmspawn/vmspawn.c]: "shut down"

File: src/vmspawn/vmspawn.c [hunk 10]
```diff
@@ -2029,9 +2092,21 @@ static int run_virtual_machine(int kvm_device_fd, int vhost_device_fd) {
...
+                ssh_info = (SSHInfo) {
+                        .cid = child_cid,
+                        .private_key_path = ssh_private_key_path,
+                        .port = 22,
+                };
+
+                (void) sd_event_add_signal(event, NULL, SIGINT | SD_EVENT_SIGNAL_PROCMASK, forward_signal_to_vm_pid1, &ssh_info);
+                (void) sd_event_add_signal(event, NULL, SIGTERM | SD_EVENT_SIGNAL_PROCMASK, forward_signal_to_vm_pid1, &ssh_info);
+        } else {
+                /* if we don't know the SSH information fallback to signalling qemu */
```
Comment (keszybz) [inline src/vmspawn/vmspawn.c]: I don't grok this sentence.

File: src/vmspawn/vmspawn.c [hunk 11]
```diff
@@ -1257,6 +1319,7 @@ static int run_virtual_machine(int kvm_device_fd, int vhost_device_fd) {
         _cleanup_close_ int notify_sock_fd = -EBADF;
         _cleanup_strv_free_ char **cmdline = NULL;
         _cleanup_free_ int *pass_fds = NULL;
+        SSHInfo ssh_info;
```
Comment (keszybz) [inline src/vmspawn/vmspawn.c]: Maybe add a comment that this is only used conditionally, but must survive until the function finishes? I think it'd be better to even move this before definition of `bus`, to underline the fact that it needs to survive longer.

**Review by keszybz — CHANGES_REQUESTED** [review]
Since we're supposed to merge this after -rc1, the commits must include a solid explanation what is being done and why and what the implications are.

**Comment by keszybz** [comment]
Code LGTM.

I think it's fine to merge this now. The first commit is a fix, and the other two are relatively simple plumbing changes that will make `machinectl shell` much more useful. The work of vmspawn and related parts is a big part of this release, so we want to make it work as well as we can. We're after -rc1, but I think this a change of this magnitude is reasonable if it complements some changes that were done earlier in the cycle.

**Comment by keszybz** [comment]
Ubunutautopkgtests are busted.

**Comment by evverx** [comment]
Looking at
```sh
ssh: Could not resolve hostname vsock/2747562528: Name or service not known
Failed to get init process of VM: Transport endpoint is not connected
```
I'm not sure stuff should get stuck like that for several hours but OK.

**Pull request merged**